tests/integration/test_schema.py

```diff
@@ -2903,3 +2947,266 @@ async def test_invalid_schema_should_provide_good_error_messages(registry_async_
+    progresses to write new version of schema. When getting the ID for the schema the full dictionary of schemas
+    was iterated and match found and reused schema id returned. Version is incremented.
+
+    Test needs to inject the invalid data to schemas topic and start Karapace for startup data ingestion. The duplicate
+    ids are checked only on startup.
+
+    Before the fix the full list of schemas after receiving all records from schemas topic:
+     * id: 2, schema: {"type":"string"}
+    After duplicate id skipping the full list of schemas after receiving all records from schemas topic:
+     * id: 2, schema: "string"
+
+    The resulting state in SchemaReader after the test is:
+    * Schemas:
+      * id: 2, schema: "string"
+    * Subjects:
+      * subject_first_subject_record_with_good_id
+        * id: 2, ver: 1, schema: "string"
+      * subject_second_subject_record_with_good_id
+        * id: 2, ver: 1, schema: "string"
+      * subject_third_subject_record_with_duplicate_id
+        * id: 2, ver: 1, schema: "string"
+
+    * Hash pointers to id (key is subject/schema):
+      * id: subject_first_subject_record_with_good_id/"string", id: 2
+      * id: subject_second_subject_record_with_good_id/"string", id: 2
+      * id: subject_first_subject_record_with_good_id/{"type": "string"}, id: 2
+      * id: subject_third_subject_record_with_duplicate_id/"string", id: 2
+    """
+
+    def _clear_test_name(name: str) -> str:
+        # Based on:
+        # https://github.com/pytest-dev/pytest/blob/238b25ffa9d4acbc7072ac3dd6d8240765643aed/src/_pytest/tmpdir.py#L189-L194
+        # The purpose is to return a similar name to make finding matching logs easier
+        return re.sub(r"[\W]", "_", name)[:30]
+
+    # If registry_url is defined the test is run to external registry.
+    # Implies also the use of external Kafka by kafka-bootstrap-servers test parameter.
+    registry_url = request.config.getoption("registry_url")
+    schemas_topic = new_random_name("_schemas") if not registry_url else "_schemas"
+    config = {"bootstrap_uri": kafka_servers.bootstrap_servers, "topic_name": schemas_topic}
+
+    if not registry_url:
+        admin_client = KafkaAdminClient(
+            bootstrap_servers=kafka_servers.bootstrap_servers,
+        )
+        admin_client.create_topics(
+            [
+                NewTopic(
+                    name=schemas_topic,
+                    num_partitions=1,
+                    replication_factor=1,
+                    topic_configs={"cleanup.policy": "compact"},
+                )
+            ],
+            timeout_ms=1000,
+        )
+        admin_client.close()
+
+    first_subject = create_subject_name_factory("first_subject_record_with_good_id")()
+    first_message_key = {
+        "keytype": "SCHEMA",
+        "subject": first_subject,
+        "version": 1,
+        "magic": 1,
+    }
+    first_schema = '"string"'
```

Comment:
Maybe renaming `first_schema` to something like `primitive_schema` so it is comparable to `schema_with type`? Also readability would be improved if both schemas are introduced first next to each other.

Reply:
The issue here is not the primitive but the order of schemas, therefore I named as first and second etc.